File: examples/rust/src/bin/mdi_command.rs
//! MDI Command Example
//!
//! Execute G-code commands via MDI (Manual Data Input) mode.
//! This is useful for sending individual G-code commands without loading a file.
//!
//! # Usage
//!
//! ```bash
//! cargo run --bin mdi_command -- "G0 X10 Y10"
//! cargo run --bin mdi_command -- --interactive
//! ```
//!
//! # Safety Warning
//!
//! MDI commands execute immediately on the machine. Understand what
//! each command does before running it.

use clap::Parser;
use linuxcnc_grpc::linuxcnc::linux_cnc_service_client::LinuxCncServiceClient;
use linuxcnc_grpc::linuxcnc::{
    linux_cnc_command, GetStatusRequest, LinuxCncCommand, LinuxCncStatus, MdiCommand, ModeCommand,
    RcsStatus, StateCommand, TaskMode, TaskState, WaitCompleteRequest,
};
use std::io::{self, BufRead, Write};
use std::sync::atomic::{AtomicI32, Ordering};
use std::time::Duration;

#[derive(Parser, Debug)]
#[command(name = "mdi_command")]
#[command(about = "Execute MDI G-code commands")]
struct Args {
    /// gRPC server host
    #[arg(long, default_value = "localhost")]
    host: String,

    /// gRPC server port
    #[arg(long, default_value_t = 50051)]
    port: u16,

    /// Enter interactive MDI mode
    #[arg(long, short)]
    interactive: bool,

    /// Don't wait for command completion
    #[arg(long)]
    no_wait: bool,

    /// G-code command to execute
    command: Option<String>,
}

struct LinuxCncClient {
    client: LinuxCncServiceClient<tonic::transport::Channel>,
    serial: AtomicI32,
}

impl LinuxCncClient {
    async fn connect(addr: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let client = LinuxCncServiceClient::connect(addr.to_string()).await?;
        Ok(Self {
            client,
            serial: AtomicI32::new(0),
        })
    }

    fn next_serial(&self) -> i32 {
        self.serial.fetch_add(1, Ordering::SeqCst) + 1
    }

    async fn get_status(&mut self) -> Result<LinuxCncStatus, Box<dyn std::error::Error>> {
        let request = tonic::Request::new(GetStatusRequest {});
        let response = self.client.get_status(request).await?;
        Ok(response.into_inner())
    }

    async fn send_command(
        &mut self,
        command: linux_cnc_command::Command,
    ) -> Result<(RcsStatus, i32), Box<dyn std::error::Error>> {
        let serial = self.next_serial();
        let cmd = LinuxCncCommand {
            serial,
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)?
                .as_nanos() as i64,
            command: Some(command),
        };
        let response = self.client.send_command(tonic::Request::new(cmd)).await?;
        let resp = response.into_inner();
        if resp.status == RcsStatus::RcsError as i32 {
            return Err(resp.error_message.into());
        }
        Ok((
            RcsStatus::try_from(resp.status).unwrap_or(RcsStatus::RcsDone),
            serial,
        ))
    }

    async fn wait_complete(
        &mut self,
        serial: i32,
        timeout: f64,
    ) -> Result<RcsStatus, Box<dyn std::error::Error>> {
        let request = tonic::Request::new(WaitCompleteRequest { serial, timeout });
        let response = self.client.wait_complete(request).await?;
        let resp = response.into_inner();
        if resp.status == RcsStatus::RcsError as i32 {
            return Err(resp.error_message.into());
        }
        Ok(RcsStatus::try_from(resp.status).unwrap_or(RcsStatus::RcsDone))
    }

    async fn set_mode(&mut self, mode: TaskMode) -> Result<RcsStatus, Box<dyn std::error::Error>> {
        let (status, _) = self
            .send_command(linux_cnc_command::Command::Mode(ModeCommand {
                mode: mode as i32,
            }))
            .await?;
        Ok(status)
    }

    async fn set_state(
        &mut self,
        state: TaskState,
    ) -> Result<RcsStatus, Box<dyn std::error::Error>> {
        let (status, _) = self
            .send_command(linux_cnc_command::Command::State(StateCommand {
                state: state as i32,
            }))
            .await?;
        Ok(status)
    }

    async fn mdi(&mut self, gcode: &str) -> Result<(RcsStatus, i32), Box<dyn std::error::Error>> {
        self.send_command(linux_cnc_command::Command::Mdi(MdiCommand {
            command: gcode.to_string(),
        }))
        .await
    }
}

async fn ensure_mdi_ready(client: &mut LinuxCncClient) -> Result<(), Box<dyn std::error::Error>> {
    let status = client.get_status().await?;

    // Check E-stop
    if let Some(task) = &status.task {
        if task.task_state == TaskState::StateEstop as i32 {
            println!("Machine is in E-stop. Resetting...");
            client.set_state(TaskState::StateEstopReset).await?;
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
    }

    // Power on
    let status = client.get_status().await?;
    if let Some(task) = &status.task {
        if task.task_state != TaskState::StateOn as i32 {
            println!("Powering on machine...");
            client.set_state(TaskState::StateOn).await?;
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
    }

    // Set MDI mode
    let status = client.get_status().await?;
    if let Some(task) = &status.task {
        if task.task_mode != TaskMode::ModeMdi as i32 {
            println!("Setting MDI mode...");
            client.set_mode(TaskMode::ModeMdi).await?;
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
    }

    Ok(())
}

async fn execute_mdi(
    client: &mut LinuxCncClient,
    gcode: &str,
    wait: bool,
) -> Result<(), Box<dyn std::error::Error>> {
    println!("Executing: {}", gcode);

    let (status, serial) = client.mdi(gcode).await?;
    if status == RcsStatus::RcsError {
        return Err("MDI command failed".into());
    }

    if wait {
        println!("  Waiting for completion...");
        let status = client.wait_complete(serial, 60.0).await?;
        if status == RcsStatus::RcsError {
            return Err("Error during execution".into());
        }
        println!("  Done.");
    }

    Ok(())
}

async fn interactive_mode(client: &mut LinuxCncClient) -> Result<(), Box<dyn std::error::Error>> {
    println!("\nInteractive MDI Mode");
    println!("Type G-code commands to execute. Type 'quit' or 'exit' to quit.");
    println!("Type 'status' to show current position.\n");

    let stdin = io::stdin();
    let handle = stdin.lock();

    for line in handle.lines() {
        print!("MDI> ");
        io::stdout().flush()?;

        let cmd = line?.trim().to_string();
        if cmd.is_empty() {
            continue;
        }

        let lower = cmd.to_lowercase();
        if lower == "quit" || lower == "exit" || lower == "q" {
            break;
        }

        if lower == "status" {
            let status = client.get_status().await?;
            if let Some(pos) = status.position.as_ref().and_then(|p| p.actual_position.as_ref()) {
                println!("Position: X={:.4} Y={:.4} Z={:.4}", pos.x, pos.y, pos.z);
            }
            continue;
        }

        if lower == "help" {
            println!("Commands:");
            println!("  <G-code>  - Execute G-code command");
            println!("  status    - Show current position");
            println!("  quit      - Exit interactive mode");
            continue;
        }

        // Ensure we're still in MDI mode
        let status = client.get_status().await?;
        if let Some(task) = &status.task {
            if task.task_mode != TaskMode::ModeMdi as i32 {
                if let Err(e) = ensure_mdi_ready(client).await {
                    eprintln!("Failed to re-enter MDI mode: {}", e);
                    continue;
                }
            }
        }

        if let Err(e) = execute_mdi(client, &cmd, true).await {
            eprintln!("Error: {}", e);
        }
    }

    Ok(())
}

fn print_usage() {
    println!("Usage: mdi_command [OPTIONS] [COMMAND]");
    println!();
    println!("Examples:");
    println!("  mdi_command \"G0 X10 Y10\"");
    println!("  mdi_command \"G1 X20 F100\"");
    println!("  mdi_command --interactive");
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();

    if args.command.is_none() && !args.interactive {
        print_usage();
        std::process::exit(1);
    }

    let addr = format!("http://{}:{}", args.host, args.port);
    let mut client = LinuxCncClient::connect(&addr).await?;

    // Ensure machine is ready for MDI
    ensure_mdi_ready(&mut client).await?;

    if args.interactive {
        interactive_mode(&mut client).await?;
    } else if let Some(command) = args.command {
        execute_mdi(&mut client, &command, !args.no_wait).await?;

        // Show final position
        let status = client.get_status().await?;
        if let Some(pos) = status.position.as_ref().and_then(|p| p.actual_position.as_ref()) {
            println!("Position: X={:.4} Y={:.4} Z={:.4}", pos.x, pos.y, pos.z);
        }
    }

    Ok(())
}

File: examples/node/mdi_command.ts
/**
 * MDI Command Example
 *
 * Execute G-code commands via MDI (Manual Data Input) mode.
 * This is useful for sending individual G-code commands without loading a file.
 *
 * Usage:
 *   npx tsx mdi_command.ts "G0 X10 Y10"
 *   npx tsx mdi_command.ts --interactive
 *
 * Safety Warning:
 *   MDI commands execute immediately on the machine. Understand what
 *   each command does before running it.
 */

import * as readline from "readline";
import { Metadata } from "@grpc/grpc-js";
import { program } from "commander";
import {
  LinuxCNCServiceClient,
  GetStatusRequest,
  LinuxCNCCommand,
  WaitCompleteRequest,
  LinuxCNCStatus,
  CommandResponse,
  TaskMode,
  TaskState,
  RcsStatus,
  credentials,
} from "linuxcnc-grpc";

program
  .argument("[command]", "G-code command to execute")
  .option("--host <host>", "gRPC server host", "localhost")
  .option("--port <port>", "gRPC server port", "50051")
  .option("-i, --interactive", "Enter interactive MDI mode")
  .option("--no-wait", "Don't wait for command completion")
  .parse();

const opts = program.opts();
const args = program.args;
const command = args[0];
const address = `${opts.host}:${opts.port}`;

if (!command && !opts.interactive) {
  console.log("Usage: npx tsx mdi_command.ts [options] \"G-code command\"");
  console.log("       npx tsx mdi_command.ts --interactive");
  console.log("\nExamples:");
  console.log('  npx tsx mdi_command.ts "G0 X10 Y10"');
  console.log('  npx tsx mdi_command.ts "G1 X20 F100"');
  console.log("  npx tsx mdi_command.ts --interactive");
  process.exit(1);
}

console.error(`Connecting to ${address}...`);
const client = new LinuxCNCServiceClient(
  address,
  credentials.createInsecure(),
  {
    'grpc.initial_reconnect_backoff_ms': 1000,
    'grpc.max_reconnect_backoff_ms': 5000,
  }
);

const RPC_DEADLINE = 5000;

let serial = 0;

function nextSerial(): number {
  return ++serial;
}

function getStatus(): Promise<LinuxCNCStatus> {
  return new Promise((resolve, reject) => {
    client.getStatus(GetStatusRequest.create(), new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, status) => {
      if (err) reject(err);
      else resolve(status);
    });
  });
}

function sendCommand(cmd: LinuxCNCCommand): Promise<CommandResponse> {
  cmd.serial = nextSerial();
  cmd.timestamp = Date.now() * 1000000;
  return new Promise((resolve, reject) => {
    client.sendCommand(cmd, new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, response) => {
      if (err) reject(err);
      else resolve(response);
    });
  });
}

function waitComplete(cmdSerial: number, timeout: number): Promise<CommandResponse> {
  const request = WaitCompleteRequest.create();
  request.serial = cmdSerial;
  request.timeout = timeout;
  const waitDeadline = new Date(Date.now() + (timeout * 1000) + 5000);
  return new Promise((resolve, reject) => {
    client.waitComplete(request, new Metadata(), { deadline: waitDeadline }, (err, response) => {
      if (err) reject(err);
      else resolve(response);
    });
  });
}

function setMode(mode: TaskMode): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.mode = { mode };
  return sendCommand(cmd);
}

function setState(state: TaskState): Promise<CommandResponse> {
  const cmd = LinuxCNCCommand.create();
  cmd.state = { state };
  return sendCommand(cmd);
}

async function mdi(gcode: string): Promise<{ response: CommandResponse; serial: number }> {
  const cmd = LinuxCNCCommand.create();
  cmd.mdi = { command: gcode };
  const response = await sendCommand(cmd);
  return { response, serial: cmd.serial };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function ensureMDIReady(): Promise<boolean> {
  let status = await getStatus();

  // Check E-stop
  let task = status.task;
  if (!task) {
    console.log("No task status available");
    return false;
  }
  if (task.taskState === TaskState.STATE_ESTOP) {
    console.log("Machine is in E-stop. Resetting...");
    const resp = await setState(TaskState.STATE_ESTOP_RESET);
    if (resp.status !== RcsStatus.RCS_DONE) {
      console.log(`Failed to reset E-stop: ${resp.errorMessage}`);
      return false;
    }
    await sleep(100);
  }

  // Power on
  status = await getStatus();
  task = status.task;
  if (!task) {
    console.log("No task status available");
    return false;
  }
  if (task.taskState !== TaskState.STATE_ON) {
    console.log("Powering on machine...");
    const resp = await setState(TaskState.STATE_ON);
    if (resp.status !== RcsStatus.RCS_DONE) {
      console.log(`Failed to power on: ${resp.errorMessage}`);
      return false;
    }
    await sleep(100);
  }

  // Set MDI mode
  status = await getStatus();
  task = status.task;
  if (!task) {
    console.log("No task status available");
    return false;
  }
  if (task.taskMode !== TaskMode.MODE_MDI) {
    console.log("Setting MDI mode...");
    const resp = await setMode(TaskMode.MODE_MDI);
    if (resp.status !== RcsStatus.RCS_DONE) {
      console.log(`Failed to set MDI mode: ${resp.errorMessage}`);
      return false;
    }
    await sleep(100);
  }

  return true;
}

async function executeMDI(gcode: string, wait: boolean): Promise<boolean> {
  console.log(`Executing: ${gcode}`);

  const { response, serial } = await mdi(gcode);
  if (response.status === RcsStatus.RCS_ERROR) {
    console.log(`  Error: ${response.errorMessage}`);
    return false;
  }

  if (wait) {
    console.log("  Waiting for completion...");
    const resp = await waitComplete(serial, 60.0);
    if (resp.status === RcsStatus.RCS_ERROR) {
      console.log(`  Error during execution: ${resp.errorMessage}`);
      return false;
    }
    console.log("  Done.");
  }

  return true;
}

async function interactiveMode(): Promise<void> {
  console.log("\nInteractive MDI Mode");
  console.log("Type G-code commands to execute. Type 'quit' or 'exit' to quit.");
  console.log("Type 'status' to show current position.\n");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const prompt = (): void => {
    rl.question("MDI> ", async (input) => {
      const cmd = input.trim();

      if (!cmd) {
        prompt();
        return;
      }

      const lower = cmd.toLowerCase();
      if (lower === "quit" || lower === "exit" || lower === "q") {
        rl.close();
        return;
      }

      if (lower === "status") {
        const status = await getStatus();
        const position = status.position;
        if (!position || !position.actualPosition) {
          console.log("No position data available");
          prompt();
          return;
        }
        const pos = position.actualPosition;
        console.log(`Position: X=${pos.x.toFixed(4)} Y=${pos.y.toFixed(4)} Z=${pos.z.toFixed(4)}`);
        prompt();
        return;
      }

      if (lower === "help") {
        console.log("Commands:");
        console.log("  <G-code>  - Execute G-code command");
        console.log("  status    - Show current position");
        console.log("  quit      - Exit interactive mode");
        prompt();
        return;
      }

      try {
        // Ensure we're still in MDI mode
        const status = await getStatus();
        const task = status.task;
        if (task && task.taskMode !== TaskMode.MODE_MDI) {
          if (!(await ensureMDIReady())) {
            console.log("Failed to re-enter MDI mode");
            prompt();
            return;
          }
        }

        await executeMDI(cmd, true);
      } catch (err) {
        console.log(`Error: ${err}`);
      }

      prompt();
    });
  };

  rl.on("close", () => {
    console.log("\nExiting...");
    client.close();
    process.exit(0);
  });

  prompt();
}

async function main(): Promise<void> {
  try {
    // Ensure machine is ready for MDI
    if (!(await ensureMDIReady())) {
      console.log("Could not prepare machine for MDI");
      process.exit(1);
    }

    if (opts.interactive) {
      await interactiveMode();
    } else {
      const success = await executeMDI(command, opts.wait !== false);
      if (!success) {
        process.exit(1);
      }

      // Show final position
      const status = await getStatus();
      const position = status.position;
      if (!position || !position.actualPosition) {
        console.error("No position data available");
      } else {
        const pos = position.actualPosition;
        console.log(`Position: X=${pos.x.toFixed(4)} Y=${pos.y.toFixed(4)} Z=${pos.z.toFixed(4)}`);
      }
      client.close();
    }
  } catch (err) {
    console.error(`Error: ${err}`);
    client.close();
    process.exit(1);
  }
}

main();
